# SCRIPT
Single-cell Cis-regulatory Relationship Identifier based on Pre-Trained graph attention networks

## Overview
Single-cell *cis*-regulatory relationships (CRRs) are crucial for deciphering transcriptional regulation and elucidating the pathogenic mechanisms of disease-associated non-coding variants. Existing methods fail to incorporate causal biological principles to model whole-genome cis-regulatory mechanisms, resulting in their insufficient performance on single-cell CRRs. Here, we present SCRIPT (Single-cell Cis-regulatory Relationship Identifier based on Pre-Trained graph attention networks), a unified computational framework with two key innovations. First, SCRIPT utilizes graph casual attention networks to simulate transcriptional regulation, leveraging empirical CRR evidence to ensure biologically grounded modeling. Second, it employs a graph autoencoder for efficient representation learning, pretrained on atlas-scale scATAC-seq data to establish a foundational understanding of cis-regulation. SCRIPT significantly outperforms two state-of-the-art methods with a mean AUC of 0.9, and achieves a sixfold relative increase in predicting long-range CRRs. Furthermore, SCRIPT is applied to reveal the pathogenic mechanisms underlying the complex-disease-associated variants, providing a promising avenue for genetic diagnosis and drug target discovery.

## Installation

We recommend installing SCRIPT within a virtual environment. For Conda, use the following command to create and activate a dedicated environment:

```shell
conda create -n script python=3.10.6 && source activate script
```

Next, install the required dependencies with:

```shell
pip install torch==2.2.1 torchvision==0.17.1 torchaudio==2.2.1 --index-url https://download.pytorch.org/whl/cu121
pip install dgl -f https://data.dgl.ai/wheels/torch-2.2/cu121/repo.html
pip install -r requirements.txt
```

## Tutorial

The useage of SCRIPT contains three steps.

The first step is generation of training data. Users need to provide a pair of scATAC-seq and scRNA-seq datasets. The two datasets are from the same tissue, and their cell types need to be aligned.

You can run the following code to perform the first step:

```
python generate_training_data.py --file_atac ./example_data/input_data/atac_cortex.h5ad --file_rna ./example_data/input_data/RNA_countsum.txt --path_out ./example_data/output_data
```
The second step is the simulation of cis-transcriptional regulation. Users need to input the training data generated by the first step.

You can run the following code to perform the second step:

```
python transcription_simulation.py --path_in ./example_data/output_data
```
The third step is the calculation of single-cell regulation scores by attribution method. Users need to input the training data generated by the first step, the configures and paremeters of the simulation model trained in the second step. 

You can run the following code to perform the third step:

```
python attribution.py --path_in ./example_data/output_data --path_out ./example_data/res_attr --config_simulation ./checkpoints/[PID_num]/args.pkl --model_simulation ./checkpoints/[PID_num]/best_model.pt
```
Single-cell regulation scores are saved in the '"./example_data/res_attr" folder.
